frc46_token/src/token/state.rs

```diff
@@ -506,114 +511,169 @@ impl TokenState {
     /// Checks that there are no zero balances, zero allowances or empty allowance maps explicitly
     /// stored in the blockstore. Checks that balances, total supply, allowances are never negative.
     /// Checks that sum of all balances matches total_supply. Checks that no allowances are stored
-    /// where operator == owner.
+    /// where operator == owner. Checks that all balances are a multiple of the granularity.
+    ///
+    /// Returns a state summary that can be used to check application specific invariants.
     pub fn check_invariants<BS: Blockstore>(
         &self,
         bs: &BS,
-    ) -> std::result::Result<(), StateInvariantError> {
+        granularity: u64,
+    ) -> std::result::Result<StateSummary, Vec<StateInvariantError>> {
+        // accumulate errors encountered in the state
+        let mut errors: Vec<StateInvariantError> = vec![];
+
+        // helper to decode addresses
+        let decode_key_addr =
+            |key: &BytesKey, errors: &mut Vec<StateInvariantError>| -> Option<ActorID> {
+                match decode_actor_id(key) {
+                    Some(actor_id) => Some(actor_id),
+                    None => {
+                        errors.push(StateInvariantError::InvalidBytesKey(key.clone()));
+                        None
+                    }
+                }
+            };
+
         // check total supply
         if self.supply.is_negative() {
-            return Err(StateInvariantError::SupplyNegative(self.supply.clone()));
+            errors.push(StateInvariantError::SupplyNegative(self.supply.clone()));
         }
 
         // check balances
         let mut balance_sum = TokenAmount::zero();
-        let mut maybe_err: Option<StateInvariantError> = None;
-        let balances = self.get_balance_map(bs)?;
-        let res = balances.for_each(|owner_key, balance| {
-            let owner = match decode_actor_id(owner_key) {
-                None => {
-                    bail!("invalid key in balances map")
-                }
-                Some(a) => a,
-            };
-            // all balances must be positive
-            if balance.is_negative() {
-                maybe_err = Some(StateInvariantError::BalanceNegative {
-                    account: owner,
-                    balance: balance.clone(),
-                });
-                bail!("invariant failed")
-            }
-            // zero balances should not be stored in the Hamt
-            if balance.is_zero() {
-                maybe_err = Some(StateInvariantError::ExplicitZeroBalance(owner));
-                bail!("invariant failed")
+        let balances = match self.get_balance_map(bs) {
+            Ok(map) => map,
+            Err(e) => {
+                errors.push(StateInvariantError::State(e));
+                return Err(errors);
```

Comment:
Move the body of the balance iteration inside the Ok branch so that you can continue checking other parts of the state even if balances can't be loaded.